### PR TITLE
[ROCm] MIOpen integration, more tests enabled, bug fixes

### PR DESCRIPTION
**Pull request by iotamudelta**
* first integration of MIOpen for batch norm and conv on ROCm
* workaround a ROCm compiler bug exposed by elementwise_kernel through explicit capture of variables in the densest packing
* workaround a ROCm compiler bug exposed by having `extern "C" __host__` as a definition and just `__host__` in the implementation through the hipify script
* use fabs() in accordance with C++11 for double absolute, not ::abs() which is integer-only on ROCm
* enable test_sparse set on CI, skip tests that don't work currently on ROCm
* enable more tests in test_optim after the elementwise_bug got fixed
* enable more tests in test_dataloader
* improvements to hipification and ROCm build

With this, resnet18 on CIFAR data trains without hang or crash in our tests.

### CODE REVIEW

**Comment by iotamudelta** [comment]
With fix landed:

@pytorchbot retest this please

File: aten/src/ATen/cudnn/Handle.cpp [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#include "Handles.h"
```
Comment (Jorghi12) [inline aten/src/ATen/cudnn/Handle.cpp]: Why are we changing the name of the file from "Handles.cpp" to "Handle.cpp" ?

File: aten/src/ATen/cudnn/Handle.h [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#pragma once
```
Comment (Jorghi12) [inline aten/src/ATen/cudnn/Handle.h]: Why are we changing the name of the file from "Handles.h" to "Handle.h" ?

File: aten/src/ATen/cudnn/Handle.cpp (hunk 1 above)
Comment (iotamudelta) [reply aten/src/ATen/cudnn/Handle.cpp]: By request from @ezyang

File: aten/src/ATen/cudnn/Handle.h (hunk 2 above)
Comment (iotamudelta) [reply aten/src/ATen/cudnn/Handle.h]: By request from @ezyang

File: aten/src/ATen/cudnn/Descriptors.h [hunk 3]
```diff
@@ -339,7 +339,7 @@ union Constant
   double d;
   Constant(cudnnDataType_t dataType, double value) {
     if (dataType == CUDNN_DATA_HALF || dataType == CUDNN_DATA_FLOAT) {
-      f = (float) value;
+      f = static_cast<float>(value);
```
Comment (Jorghi12) [inline aten/src/ATen/cudnn/Descriptors.h]: Curious, why is the static_cast necessary here?

Comment (iotamudelta) [reply aten/src/ATen/cudnn/Descriptors.h]: Simply better to use the explicit C++ syntax over C-style cast, I believe.

File: aten/src/THC/THCNumerics.cuh [hunk 4]
```diff
@@ -663,7 +663,7 @@ struct THCNumerics<double> {
   static inline __host__ __device__  double tanh (double a) { return  ::tanh(a); }
   static inline __host__ __device__  double erf  (double a) { return   ::erf(a); }
   static inline __host__ __device__  double erfc (double a) { return  ::erfc(a); }
-  static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
+  static inline __host__ __device__  double abs  (double a) { return   fabs(a); }
```
Comment (ezyang) [inline aten/src/THC/THCNumerics.cuh]: This should have a comment saying why fabs is needed here.

File: tools/amd_build/pyHIPIFY/hipify-python.py [hunk 5]
```diff
@@ -430,8 +430,17 @@ def find_kernel_bounds(string):
 def find_closure_group(input_string, start, group):
     """Generalization for finding a balancing closure group
 
-    e.g. if group = ["(", ")"], then finds the first balanced parantheses.
+         if group = ["(", ")"], then finds the first balanced parantheses.
```
Comment (ezyang) [inline tools/amd_build/pyHIPIFY/hipify-python.py]: nit: parentheses

**Review by ezyang — APPROVED** [review]

File: aten/src/THC/THCNumerics.cuh (hunk 4 above)
Comment (Jorghi12) [inline aten/src/THC/THCNumerics.cuh]: Probably not the greatest idea to reduce the precision from the double overload of abs to fabs. It's better to add the implementation for the abs double overload directly in the HIP stack in order to maintain parallelism here in the code.

File: test/cpp_extensions/cudnn_extension.cpp [hunk 6]
```diff
@@ -14,7 +14,7 @@
 
 #include <ATen/cudnn/Descriptors.h> // for TensorDescriptor
 #include <ATen/cuda/Exceptions.h> // for CUDNN_CHECK
-#include <ATen/cudnn/Handles.h> // for getCudnnHandle
```
Comment (Jorghi12) [inline test/cpp_extensions/cudnn_extension.cpp]: The name changes for the files we already have are unnecessary.

File: tools/amd_build/disabled_features.yaml [hunk 7]
```diff
@@ -78,7 +72,9 @@
       {
         "path": "aten/src/THC/THCTensorRandom.cu",
         "s_constants": {
-          "struct curandStateMtgp32*": "curandStateMtgp32*"
+          "struct curandStateMtgp32*": "curandStateMtgp32*",
+          "__host__ void THCRandom_getRNGState": "extern \"C\" __host__ void THCRandom_getRNGState",
```
Comment (Jorghi12) [inline tools/amd_build/disabled_features.yaml]: Could you comment with an explanation for why this extra modification is necessary? The declarations of these functions already have "extern C" (via THC_API macro) inside aten/src/THC/THCTensorRandom.h. Thus these functions will already have support for linkage to modules written in the C programming language. It seems redundant to me.

The C++ standard:
_A function can be re-declared without a linkage specification after it was declared with a language specification, the second declaration will reuse the first language linkage._ 

https://en.cppreference.com/w/cpp/language/language_linkage

File: aten/src/THC/THCNumerics.cuh (hunk 4 above)
Comment (iotamudelta) [reply aten/src/THC/THCNumerics.cuh]: It does not reduce precision. http://www.cplusplus.com/reference/cmath/fabs/

File: test/cpp_extensions/cudnn_extension.cpp (hunk 6 above)
Comment (iotamudelta) [reply test/cpp_extensions/cudnn_extension.cpp]: please discuss w/ @ezyang who requested this

File: tools/amd_build/disabled_features.yaml (hunk 7 above)
Comment (iotamudelta) [reply tools/amd_build/disabled_features.yaml]: As in the commit message: this is a workaround for a compiler frontend bug we filed already.

File: tools/amd_build/pyHIPIFY/hipify-python.py (hunk 5 above)
Comment (iotamudelta) [reply tools/amd_build/pyHIPIFY/hipify-python.py]: but isn't plural correct here?

File: tools/amd_build/disabled_features.yaml (hunk 7 above)
Comment (Jorghi12) [reply tools/amd_build/disabled_features.yaml]: Just saw, awesome!

**Comment by ezyang** [comment]
ROCm failures look legit?

```
0:57:50 Generating AMD GCN kernel failed in llc for target: gfx900
00:57:50 /opt/rocm/hcc/bin/llc: /tmp/tmp.0PiILY6VCc/caffe2_hip_generated_BatchNorm.cpp.kernel.bc-gfx900.isabin.opt.bc: error: Could not open input file: No such file or directory
00:57:50 clang-7.0_original: error: linker command failed with exit code 1 (use -v to see invocation)
00:57:50 Died at /opt/rocm/hip/bin/hipcc line 565.
00:57:50 caffe2/CMakeFiles/caffe2_hip.dir/build.make:2248: recipe for target 'lib/libcaffe2_hip.so' failed
```

Also note to fb people: this diff needs fbcode changes

**Comment by yf225** [comment]
This PR seems to be causing `py2-clang3.8-rocm1.7.1-ubuntu16.04-test` to fail on master:

```
23:57:06 test_factory_dense_dims (__main__.TestCudaSparse) ... Traceback (most recent call last):
23:57:06   File "test/run_test.py", line 363, in <module>
23:57:06     main()
23:57:06   File "test/run_test.py", line 355, in main
23:57:06     raise RuntimeError(message)
23:57:06 RuntimeError: test_sparse failed! Received signal: SIGSEGV
```

**Pull request closed without merging**